downstairs/src/region.rs

```diff
@@ -1053,6 +1014,136 @@ impl Region {
+    /// region is backed by a ZFS dataset. Issue a `syncfs` syscall, which
+    /// will result in a `zfs_sync` to the entire region dataset.
+    ///
+    /// The codepath is roughly as follows
+    /// - The syscall calls `fdsync(.., FDSYNC_FS)`
+    /// - `fdsync` calls `VFS_SYNCFS` for the appropriate filesystem
+    /// - For ZFS, this calls `zfs_syncfs`
+    /// - `zfs_syncfs` calls `zfs_sync(vfsp, 0, cr)`
+    /// - `zfs_sync` calls `zil_commit(zfsvfs->z_log, 0)` if `zfsvfs->z_log` is
+    ///   present; it should always be present, because it's only set to `NULL`
+    ///   during `zfsvfs_teardown`.
+    ///
+    /// Quoth the `zil_commit` docstring:
+    ///
+    /// > If "foid" is zero, this means all "synchronous" and "asynchronous"
+    /// > txs, for all objects in the dataset, will be committed to stable
+    /// > storage prior to zil_commit() returning.
+    ///
+    /// The extents all live within the same dataset, so this is what we want!
+    #[cfg(feature = "omicron-build")]
+    fn flush_extents(
+        &mut self,
+        dirty_extents: &BTreeSet<ExtentId>,
+        flush_number: u64,
+        gen_number: u64,
+        job_id: JobId,
+    ) -> Result<(), CrucibleError> {
+        #[cfg(not(target_os = "illumos"))]
+        compile_error!("syncfs-based flush is only supported on illumos");
+        use std::os::fd::AsRawFd;
+
+        // Prepare to flush extents by writing metadata to files (if needed)
+        //
+        // Some extents may not _actually_ be dirty, so we'll skip them here
+        let mut flushed_extents = vec![];
+        let log = self.log.clone();
+        for eid in dirty_extents {
+            let extent = self.get_opened_extent_mut(*eid);
+            if extent.pre_flush(flush_number, gen_number, job_id, &log)? {
+                flushed_extents.push(*eid);
+            } else {
+                warn!(
+                    self.log,
+                    "extent {eid} was in dirty_extents but not actually dirty"
+                );
+            }
+        }
+
+        // Execute the syscall
+        let dir = self.dir.clone();
+
+        // Do the flush in a worker thread if possible, to avoid blocking the
+        // main tokio thread pool.
+        run_blocking(move || -> Result<(), CrucibleError> {
+            let file = File::open(&dir)?;
+            let rc = unsafe { syncfs(file.as_raw_fd()) };
+            if rc != 0 {
+                let e = nix::errno::Errno::from_raw(rc);
+                let e: std::io::Error = e.into();
+                Err(CrucibleError::from(e))
+            } else {
+                Ok(())
+            }
+        })?;
+
+        // After the bits have been committed to durable storage, execute any
+        // post flush routine that needs to happen
+        for eid in flushed_extents {
+            let extent = self.get_opened_extent_mut(eid);
+            extent.post_flush(flush_number, gen_number, job_id)?;
```

Comment:
If we do all the flushes in one pass, then do the `post_flush` in a second loop, our DTrace counters for `cdt::extent__flush__done` will not fire till the `post_flush` loop.  I don't know if that really matters, but we should at least understand it if we are looking at per-extent flush times, as I believe the first extent we flush will include in its final `extent_flush_done` value the time for all the other extents to flush as well.

We still have the per extent `file_flush_start` and `file_flush_done` counters, which probably are the details we care about, but if we are trying to profile the "post flush work per extent" I don't think we will be gathering that info correctly.